### ISSUE TITLE
Add keyboard navigation to NumberControl

### ISSUE THREAD

**Issue by zepumph** [phetsims/scenery-phet#341]
In the keyboard navigation meeting today, we designed a prototype mockup for how the `NumberControl` component should look. We used Rutherford Scattering, Plinko Probability, and Pendulum Lab to think about what is best for the interaction. @amanda-phet could you please add the mockup to this issue so that @jessegreenberg or I can implement?

**Comment by amanda-phet** [comment]
Ruthorford scattering would need extra space to make room for the thin outer box. 
![screen shot 2017-10-24 at 12 10 03 pm](https://user-images.githubusercontent.com/7852916/31960247-65da712a-b8b4-11e7-9bcb-6359cb126712.png)
![screen shot 2017-10-24 at 12 10 15 pm](https://user-images.githubusercontent.com/7852916/31960248-65f3361a-b8b4-11e7-9f1a-f0f45776fc4d.png)
![screen shot 2017-10-24 at 12 10 23 pm](https://user-images.githubusercontent.com/7852916/31960249-6605e530-b8b4-11e7-89dc-507a4cc9ee02.png)

**Comment by zepumph** [comment]
It has been added now to master.

**Comment by zepumph** [comment]
@amanda-phet @emily-phet @arouinfar what do you guys think?
https://www.colorado.edu/physics/phet/dev/html/plinko-probability/1.2.0-dev.0/plinko-probability_en.html?accessibility

**Comment by amanda-phet** [comment]
Looks great to me! 

My one concern is if the thin outer box is slightly transparent that it won't work well on sims with a black background, or colored panel (which is often beige or gray in addition to white or black). So it would be good to see it on a few more sims. If it doesn't work well in Ruthorford Scattering, maybe we opt for a thinner but fully opaque stroke instead?

**Comment by arouinfar** [comment]
Looks good @zepumph! 

> So it would be good to see it on a few more sims. If it doesn't work well in Ruthorford Scattering, maybe we opt for a thinner but fully opaque stroke instead?

Great suggestion @amanda-phet. I'd also like to see how this look in other sims, and possibly use a fully opaque stroke if we notice any contrast issues.

**Comment by zepumph** [comment]
I'm currently converting Rutherford to use this component for keyboard nav, here is what it will look like:
![image](https://user-images.githubusercontent.com/6856943/32011786-40d14dfa-b962-11e7-8089-ac9b0779ba4d.png)

Perhaps not enough contrast. Once it is fully implemented you will be able to take it for a test.

**Comment by zepumph** [comment]
Our workaround to get this focusHighlight working yesterday isn't a good one. By making HSlider focusHighlightLayerable, we changed the bounds of all HSliders (thanks @phet-steele and @pixelzoom) for catching that). So @jessegreenberg and I will need to figure out another way to access the HSlider's "visibility" from the parent `NumberControl`. Marking for meeting to discuss.

**Comment by zepumph** [comment]
This was also a problem for that node added to NumberControl. I think we can fix that though by creating a shape that stays inside the bounds of the Component.

**Comment by pixelzoom** [comment]
Looks like you're doing something here related to the thumb. Keep in mind that HSlider supports a custom thumb via `options.thumbNode`.  So whatever you do needs to handle an thumb with arbitrary shape and size.

**Comment by pixelzoom** [comment]
@pixelzoom Outside my wheelhouse, buy a11y question related to NumberControl.  In the mockups for https://github.com/phetsims/scenery-phet/issues/341, I don’t see any examples where the arrow buttons have focus.  Is that not supported, or just not shown in the mockups?
GitHub
Add keyboard navigation to NumberControl · Issue #341 · phetsims/scenery-phet
In the keyboard navigation meeting today, we designed a prototype mockup for how the NumberControl component should look. We used Rutherford Scattering, Plinko Probability, and Pendulum Lab to thin...
 

@jessegreenberg 
The design decision was to not include the arrow buttons in focus order because they are redundant for keyboard interaction, the NumberControl as a whole is the element that receives  keyboard focus.
```

**Comment by zepumph** [comment]
Today during the a11y dev meeting @mbarlow12 @jessegreenberg and I really dove into trying to make this focus highlight implementation work well, without messing with the bounds. We settled on going back to the basics, removing `focusHighlightLayerable` and instead working to use the existing focusHighlight structure in FocusOverlay. We came up with two methods:

* First (very hacky): Take over the slider's focusHighlight from the parent `NumberControl` to manipulate it, add a node to it, and then set that new node back onto the slider (SO HACKY). From there you just have to adjust the layout to the correct parent context and it works. Here is important code from that:
```
var focusHighlight = new Node()
var numberControlFocusHighlightBorder = new FocusHighlightFromPath( /* make adjustments to offset translation because the slider's 0x0 is different than the current 0x0 in `NumberControl */);
focusHighlight.addChild( numberControlFocusHighlightBorder ); // add the big box from NumberControl
focusHighlight.addChild( slider.focusHighlight ); // Add the current focusHighlight to this new node so we don't loose it
slider.focusHighlight = focusHighlight;
```
Again, hacky, especially since layout is done in other functions, so we would have to have a custom layout for each layout function, let alone if you chose you own layout function.

* The other way we went about doing things was to instead try to use the `NumberControl` as the focusable entity, and pass events down to the HSlider's inputListeners. This way seemed a bit better.
Here is relevant code.
```js
    var focusHighlight = new Node();
    focusHighlight.addChild( numberControlFocusHighlightBorder );
    focusHighlight.addChild( slider.focusHighlight );
    this.focusHighlight = focusHighlight;
    this.addAccessibleInputListener( slider.accessibleInputListener );
```
Except that instead of just using all of the child's event listeners, we can pick and choose as part of an api (good idea @mbarlow12 )

**Comment by zepumph** [comment]
I again experimented with the second option today, here is the code:
```js
    var numberControlFocusHighlightBorder = new FocusHighlightFromNode( this, {
      outerLineWidth: 1,
      innerLineWidth: 1,
      innerStroke: FocusHighlightPath.OUTER_FOCUS_COLOR // make both the inner and outer focus color be the same
    } );

    var focusHighlight = new Node();
    focusHighlight.addChild( numberControlFocusHighlightBorder );
    focusHighlight.addChild( slider.focusHighlight );
    slider.focusHighlight.y = slider.y;
    slider.focusHighlight.x = slider.x;
    this.focusHighlight = focusHighlight;

    // add the same input listeners that the slider does, so that NumberControl keyboard interaction mimics HSlider's
    this.addAccessibleInputListener( slider.accessibleInputListenersObject );

```

It didn't seem to work though today because the layout was wrong. I think we may again have to go into the layout functions still. @jessegreenberg @mbarlow12 how did we get it to work last night without the going in the layout function? Am I remembering this right, did we even have the layout right?

**Comment by zepumph** [comment]
Maybe this is something worth looking into further:
```js
    var focusHighlight = new Node();
    focusHighlight.addChild( numberControlFocusHighlightBorder );
    focusHighlight.addChild( slider.focusHighlight );
    var sliderFocusHighlightPosition = this.globalToLocalPoint(
      slider.focusHighlight.localToGlobalPoint( new Vector2( slider.focusHighlight.x, slider.focusHighlight.y ) ) );
    slider.focusHighlight.x = sliderFocusHighlightPosition.x;
    slider.focusHighlight.y = sliderFocusHighlightPosition.y;
    this.focusHighlight = focusHighlight;

    // add the same input listeners that the slider does, so that NumberControl keyboard interaction mimics HSlider's
    this.addAccessibleInputListener( slider.accessibleInputListenersForPublicUse );

**Comment by zepumph** [comment]
@jessegreenberg @mbarlow12 and I cranked this one out today. Thanks all for help in implementing it. The meat and bones are in @jessegreenberg's commit above. I just added some pixel polishing. @emily-phet @terracoda and anyone else what do you think? NOTE: don't use keyboard nav by pressing down more than one key at a time, this is a known bug. https://www.colorado.edu/physics/phet/dev/html/rutherford-scattering/1.1.0-dev.0/?accessibility

**Comment by terracoda** [comment]
Regarding https://github.com/phetsims/scenery-phet/issues/341#issuecomment-341283975

@zepumph, @jessegreenberg  and @mbarlow12 nice work. I think it is looking pretty good. With the planned extra whitespace it should look really good.

**Comment by terracoda** [comment]
And regarding Plinko in https://github.com/phetsims/scenery-phet/issues/341#issuecomment-341283975, I think it looked really nice.

**Comment by zepumph** [comment]
>With the planned extra whitespace it should look really good.

I can hard code that pretty easily with the following line of code:
`    var numberControlFocusHighlightBorder = new FocusHighlightPath( Shape.bounds(this.bounds.setMaxY( this.bounds.bottom + 6 )), {`

But often NumberControl components are stacked tightly on top of each other. I don't want to extend it too much so that there are many cases of overlap that we have to handle downstream.

**Comment by zepumph** [comment]
- [ ] It was decided that we should not disable the tweaker buttons, but instead just make them looked 'pressed' when appropriate based on the shift + key. This will e much easier once HSlider has refactored a11y input event functions. Assigning to @jessegreenberg to reassign to me once he has completed that work.

**Comment by jessegreenberg** [comment]
@zepumph I think we are back on for this, refactoring is mostly completed in https://github.com/phetsims/sun/issues/326, waiting for code review. If you want to discuss changes let me  know, or if you would rather wait until code review is complete in that issue that is understandable too. Reassigning back to you.

**Comment by zepumph** [comment]
I got the depressed arrow look working, but I don't think it is good for master.

1. I added and `increasedEmitter` and `decreasedEmitter` to `AccessibleSlider`. 
2. Then in NumberControl (which mixes AccessibleSlider) I add listeners to them like so:
```js
    this.increasedEmitter.addListener( function() {
      rightArrowButton.buttonModel.downProperty.value = true;
      rightArrowButton.buttonModel.overProperty.value = true;

      setTimeout( function() {
        rightArrowButton.buttonModel.downProperty.value = false;
        rightArrowButton.buttonModel.overProperty.value = false;
      }, 100 );
    } );
    this.decreasedEmitter.addListener( function() {

      leftArrowButton.buttonModel.downProperty.value = true;
      leftArrowButton.buttonModel.overProperty.value = true;

      setTimeout( function() {
        leftArrowButton.buttonModel.downProperty.value = false;
        leftArrowButton.buttonModel.overProperty.value = false;
      }, 100 );

    } );
```

This is a major hack to me, as buttonModel is read/listen only. Also `setTimeout` is no good. But it worked. We could use the Joist based setTimeout, but I'm not sure about this buttonModel stuff. Honestly I'm surprised that we have never had to look into/set/listen to the visual elements of a button view before. 
@jessegreenberg do you have any other recommendations about how to proceed?

**Comment by jessegreenberg** [comment]
This feels very related to https://github.com/phetsims/sun/issues/323. I can't think of a great solution off the top of my head. For the a11y reasons listed in that issue, I have a feeling we may be stuck with something like Timer.setTimeout.

One thing we could do is add something like togglePressedState() to RectangularButtonView, so the code in NumberControl would look like
```js
    this.increasedEmitter.addListener( function() {
      arrowButton.togglePressedState();
    } );
```

and RectangularButtonView handles the timeout and the buttonModel. I believe that togglePressedState could be used for https://github.com/phetsims/sun/issues/323.

**Comment by jessegreenberg** [comment]
The above commit does something very similar to https://github.com/phetsims/scenery-phet/issues/341#issuecomment-350147488, but using phetsims/sun#323. @zepumph would you mind reviewing this commit when you return?

**Comment by zepumph** [comment]
Looks very nice! As with https://github.com/phetsims/sun/issues/318#issuecomment-370896356 I wonder if we don't want to make `increasedEmitter` more like `increasedFromA11yEmitter`.

**Comment by mbarlow12** [comment]
There may be a new bug found in https://github.com/phetsims/plinko-probability/issues/104 where mouse interaction with the thumb of the HSlider doesn't shift focus, and an extra tab is required to regain focus. As this doesn't happen in RIAW or Ohm's Law (but does for ISLC sims), I _think_ it's related to the number control implementation. @jessegreenberg should I assign myself to investigate and update?

**Comment by pixelzoom** [comment]
There has been significant work done here, but no closure, and no progress for 31 months.  So raising priority to high.

Status of this issue:

@zepumph had a question for @jessegreenberg in https://github.com/phetsims/scenery-phet/issues/341#issuecomment-370903305.

@mbarlow12 identified a potential related bug manifesting in plinko-probability in phetsims/plinko-probability#104, and asked @jessegreenberg how to proceed.

**Comment by terracoda** [comment]
We have an accessible number control in Gravity Force Lab regular and accessible number spinners in Gravity Force Lab: Basics. I don't think the design patterns have been documented in "Binder", but the PhET objects are accessible with all inputs.

**Comment by terracoda** [comment]
@pixelzoom, I think this issue stale and https://github.com/phetsims/plinko-probability/issues/104 is also likely stale since we didn't move forward with publishing that sim with Alternative Input.

**Comment by pixelzoom** [comment]
Thanks @terracoda - you're probably correct, it looks stale to me too.  The subjects of @zepumph's question in https://github.com/phetsims/scenery-phet/issues/341#issuecomment-370903305 (`increasedEmitter` and `increasedFromA11yEmitter`) don't even appear to exist anymore.   But since I'm not familiar with the work that was being done here, I'd prefer to have @jessegreenberg to wrap this up.

**Comment by jessegreenberg** [comment]
This is stale and can now be closed. The emitters referenced in https://github.com/phetsims/scenery-phet/issues/341#issuecomment-370903305 do not exist anymore, and the initial a11y code discussed in this issue has been removed or factored out in https://github.com/phetsims/scenery-phet/issues/452.
